File: p011/c/p011.c
#include <stdio.h>

#define ROWS 20
#define COLS 20
#define UNIT 4

int main(){
  
  int i, j, k, result = 0;
  int sub[4];
  int temp;
  int data[ROWS][COLS];
  FILE *fp = fopen("data.txt", "r");
  for (i = 0; i < ROWS; i++) {
    for (j = 0; j < COLS; j++) {
      fscanf(fp, "%d", &temp);
      data[i][j] = temp;
    }
  }
  fclose(fp);

  // 가로
  for (i = 0; i < ROWS; i++) {
    for (j = 0; j < COLS; j++) {

      sub[0] = 1;
      sub[1] = 1;
      sub[2] = 1;
      sub[3] = 1;

      for (k = 0; k < UNIT; k++) {
	// 가로
	if (j < COLS - UNIT) {
	  sub[0] *= data[i][j+k];
	}
	
	// 세로
	if (i < ROWS - UNIT) {
	  sub[1] *= data[i+k][j];
	}

	// 대각선
	if (i < ROWS - UNIT && j < COLS - UNIT) {
	  sub[2] *= data[i+k][j+k];
	  sub[3] *= data[i+k][j+UNIT-k];
	}
      }

      for (k = 0; k < 4; k++) {	
	if (result < sub[k]) {
	  result = sub[k];
	}
      }
    }
  }
   
  printf("%d", result);
  return 0;

}

File: p011/ruby/p011.rb
#!/usr/bin/ruby

ROWS = 20
COLS = 20
UNIT = 4

File.open("data.txt") do |f|
  data = f.readlines.map{|line|
    line.split(/\s/).map(&:to_i)
  }

  result = 0

  (0...ROWS).each do |i|
    (0...COLS).each do |j|
      sub = [1, 1, 1, 1]
      (0...UNIT).each do |k|
        sub[0] *= data[i][j+k] if j < COLS - UNIT
        sub[1] *= data[i+k][j] if i < ROWS - UNIT
        if i < ROWS - UNIT and j < COLS - UNIT
          sub[2] *= data[i+k][j+k]
          sub[3] *= data[i+k][j+UNIT-k];
        end
      end
      result = sub.max if sub.max > result
    end
  end
  p result
end

